ARTEMIS-1093 Full qualified queue name support

gaohoward opened the pull request:
Broker should support full qualified queue names (FQQN)
as well as bare queue names. This means when clients access
to a queue they have two equivalent ways to do so. One way
is by queue names and the other is by FQQN (i.e. address::qname)
names. Currently only receiving is supported.

gaohoward commented:
@clebertsuconic @mtaylor This is a resubmit of FQQN PR, with all tests fixed.

gaohoward commented:
guys please hold on this commit, I found some garbage files need to remove.

tabish121 commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpSendReceiveTest.java:
```diff
@@ -53,38 +56,61 @@
 /**
  * Test basic send and receive scenarios using only AMQP sender and receiver links.
  */
+@RunWith(Parameterized.class)
```
This seems like an unnecessary change to this test.  The intent here is to test basic AMQP protocol support not to test the FQQN support in the broker which it appears is already done in the ProtonFullQualifiedNameTest.  I'd prefer if we could keep tests targeted and not try and make them test the entirety of the broker features in one monolithic test case as that makes it harder to maintain.  If you changed this one why didn't you change every single AMQP test in the same fashion?

gaohoward replied:
@tabish121 ok, I'll remove it to keep it simple.

gaohoward commented:
OK, I removed the garbage files and reverted AmqpSendReceiveTest.

clebertsuconic commented:
Please.  While the whole testsuite ?

gaohoward commented:
There might be a problem with Jenkins machine, the failure relates to 'memory allocation'. Not related to my PR.
@clebertsuconic I ran the whole test suite on my local machine, there are few test failures but none of them related to my PR. They failed both with and without my PR. (I'll investigate those failures further when I got time).

clebertsuconic commented:
@mtaylor I will let you merge this.. but this PR is +1 from me.. 

nice job @gaohoward

clebertsuconic commented:
@gaohoward actually, the example managmeent is broken after this. Can you look please?

gaohoward commented:
@clebertsuconic sure I'll take care of it. Thanks.

gaohoward commented:
@clebertsuconic @mtaylor turns out I compared a String address with SimpleString, which always return false. Fixed that and added a simple test. 
Jenkins for the moment seems down, I'll kick off a build when it's back.

gaohoward commented:
Jenkins looking good. :)

clebertsuconic commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/util/FQQN.java:
```diff
@@ -0,0 +1,31 @@
+import org.apache.activemq.artemis.utils.CompositeAddress;
+
+public final class FQQN {
+
```
Can you remove this class, and move these two methods towards CompositeAddress

clebertsuconic replied:
Actually, I already did while merging.

gaohoward replied:
@clebertsuconic  Thanks. :)
